process/smartContract/scQueryService.go

```diff
@@ -116,60 +176,204 @@ func (service *SCQueryService) shouldAllowQueriesExecution() bool {
 
 	shouldEarlyExitBecauseOfSyncState := query.ShouldBeSynced && service.bootstrapper.GetNodeState() == common.NsNotSynchronized
 	if shouldEarlyExitBecauseOfSyncState {
-		return nil, process.ErrNodeIsNotSynced
+		return nil, nil, process.ErrNodeIsNotSynced
+	}
+
+	blockHeader, blockRootHash, err := service.extractBlockHeaderAndRootHash(query)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(blockRootHash) > 0 {
+		err = service.apiBlockChain.SetCurrentBlockHeaderAndRootHash(blockHeader, blockRootHash)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		accountsAdapter := service.blockChainHook.GetAccountsAdapter()
+		err = accountsAdapter.RecreateTrie(blockRootHash)
+		if err != nil {
+			return nil, nil, err
+		}
 	}
 
 	shouldCheckRootHashChanges := query.SameScState
 	rootHashBeforeExecution := make([]byte, 0)
 
 	if shouldCheckRootHashChanges {
-		rootHashBeforeExecution = service.blockChain.GetCurrentBlockRootHash()
+		rootHashBeforeExecution = service.mainBlockChain.GetCurrentBlockRootHash()
 	}
 
-	service.blockChainHook.SetCurrentHeader(service.blockChain.GetCurrentBlockHeader())
+	service.blockChainHook.SetCurrentHeader(service.mainBlockChain.GetCurrentBlockHeader())
 
 	service.wasmVMChangeLocker.RLock()
 	vm, _, err := scrCommon.FindVMByScAddress(service.vmContainer, query.ScAddress)
 	if err != nil {
 		service.wasmVMChangeLocker.RUnlock()
-		return nil, err
+		return nil, nil, err
 	}
 
 	query = prepareScQuery(query)
 	vmInput := service.createVMCallInput(query, gasPrice)
 	vmOutput, err := vm.RunSmartContractCall(vmInput)
 	service.wasmVMChangeLocker.RUnlock()
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	if service.hasRetriableExecutionError(vmOutput) {
 		log.Error("Retriable execution error detected. Will retry (once) executeScCall()", "returnCode", vmOutput.ReturnCode, "returnMessage", vmOutput.ReturnMessage)
 
 		vmOutput, err = vm.RunSmartContractCall(vmInput)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 	}
 
 	if query.SameScState {
 		err = service.checkForRootHashChanges(rootHashBeforeExecution)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
+		}
+	}
+
+	var blockHash []byte
+	var blockNonce uint64
+	if !check.IfNil(blockHeader) {
+		blockNonce = blockHeader.GetNonce()
+		blockHash, err = core.CalculateHash(service.marshaller, service.hasher, blockHeader)
+		if err != nil {
+			return nil, nil, err
 		}
 	}
+	blockInfo := holders.NewBlockInfo(blockHash, blockNonce, blockRootHash)
+	return vmOutput, blockInfo, nil
+}
+
+// TODO: extract duplicated code with nodeBlocks.go
+func (service *SCQueryService) extractBlockHeaderAndRootHash(query *process.SCQuery) (data.HeaderHandler, []byte, error) {
+
+	if len(query.BlockHash) > 0 {
+		blockHeader, err := service.getBlockHeaderByHash(query.BlockHash)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		header := service.getBlockHeader(blockHeader)
+		return header, header.GetRootHash(), nil
```

Comment:
I see that you extract the roothash from the following block after the 'block by hash'

I think the header to return is exactly the block by hash, but the rootHash to return here should be the ScheduledRootHash from the following block after the `block by hash`

``` go
return blockHeader, header.GetScheduledRootHash, nil
```

Reply:
updated